addons/sale/models/sale_order.py

```diff
@@ -743,6 +743,14 @@ def _onchange_company_id_warning(self):
                 }
             }
 
+    @api.onchange('company_id')
+    def _onchange_company_id(self):
+        for order in self:
+            # This can't be caught by a python constraint as it is only triggered at save and the compute method that
+            # needs this data to be set correctly before saving
+            if not order.company_id:
+                raise ValidationError(_("We can't leave this document without any company. Please select a company for this document."))
```

Comment:
Not really convinced by the phrasing of the message TBH.
`The company is required, please select one before making any other change to the sale order`

Reply:
@Feyensv I included a super call to _onchange_company_id in sale_management and changed the commit message and the ValidationError message as suggested.